### ISSUE TITLE
Handling positiveMax() in Base Modelica output

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#12418]
### Steps to Reproduce

- Open the Modelica.Fluid.Examples.Tanks.ThreeTanks model
- Flatten it with `--baseModelica`

The resulting flat code contains the `$OMC$PositiveMax()` function call (e.g., at line 1833). Besides not being a valid Modelica identifier, this function is not defined in Base Modelica, which does not have any connection primitives.

### Expected Behavior
- the function name should be a quoted-identifier, e.g. `'$OMC$PositiveMax'()`
- a function with that name should be added to the flat model package, e.g. using the implementation suggested in [Section 15.2](https://specification.modelica.org/maint/3.6/stream-connectors.html#instream-and-connection-equations) of the MLS

**Comment by perost** [comment]
How do we want to define `positiveMax`? If we use the trivial implementation suggested by the specification then we don't really need the function, we can just generate the suggested `max` call instead.

There's also a more sophisticated implementation that would need to be defined as a function, but it can't be used as is since it uses the nominal value of the flow variable. So at minimum we'd have to add that as an argument to the function.

Over all this seems like something that should maybe be a builtin operator in Base Modelica, since there isn't a single definitive implementation of `positiveMax`.

**Comment by casella** [comment]
> Over all this seems like something that should maybe be a builtin operator in Base Modelica, since there isn't a single definitive implementation of `positiveMax`.

We discussed this in the BaseModelica working group, but we agreed this may not be a good idea, as it really deals with low-level connection semantics details.

> How do we want to define `positiveMax`? If we use the trivial implementation suggested by the specification then we don't really need the function, we can just generate the suggested `max` call instead.

That's one option. However, using a custom defined function has the advantage that one could modify it when using the Base Modelica model.

> There's also a more sophisticated implementation that would need to be defined as a function, but it can't be used as is since it uses the nominal value of the flow variable. So at minimum we'd have to add that as an argument to the function.

I guess you could just constant-evaluate it and then put it as a constant in the function definition. Of course this means that we potentially need one function for each non-trivial stream connection, but I guess that's OK.

We could then define a flag to choose how you want to handle it, e.g. `--flatPositiveMax=<option>`, the default being `max` and the other option being `smooth`

@perost what do you think?

**Comment by perost** [comment]
#12429 adds a definition for `positiveMax` to Base Modelica. I only added the trivial implementation for now, but left room to add other definitions if we want.

I think it should be possible to define the smooth implementation as a function by just passing the nominal and stream values as arguments to it, but I'm not sure if the specification for it is correct. For example, why is it treating `mj.c.m_flow` as an array in `eps := relativeTolerance*min(nominal(mj.c.m_flow));`? Is it just missing the second argument of `min`?

**Comment by casella** [comment]
The specification is given as pseudo-code, not rigorous Modelica code. As I understand, `min(nominal(mj.c.m_flow))` means the minimum nominal value of all flow variables of all the connectors involved in the connection set.

**Comment by perost** [comment]
> The specification is given as pseudo-code, not rigorous Modelica code. As I understand, `min(nominal(mj.c.m_flow))` means the minimum nominal value of all flow variables of all the connectors involved in the connection set.

The beginning of the chapter says that there are assumed to be N inside connectors m<sub>j</sub>.c (j = 1:N) in a connection set, so I don't see how m<sub>j</sub>.c.m_flow could be interpreted to mean all the flow variables in the connection set. If that's really the intent, then it should be `min(nominal(mj.c.m_flow) for j in 1:N)` like used in other places.

**Comment by casella** [comment]
You are right, positiveMax is tricky, I need to consider it carefully every time I get into it 😅 

The intent is to get this mixing property always well defined. In principle, only the incoming flow matter, so one should use max(-m_j.c.m_flow, 0), but if they are all zero, then you get 0/0 which is not well defined. The trick is then to consider very small leaks (with a flow eps) instead of zero flow. 

The trivial implementation just adds small leaks everywhere, which is mostly OK as they change the mixing properties marginally as soon as there is at least one flow >> eps. However, they unnecessarily spoil the result in that case. 

The sophisticated implementation computes the s_i variable for each port i, which is the sum of the incoming flows from the other ports, and only applies the leaks when s is very small or zero. The question is: what does it mean to be very small? It of course depends on the application: a nuclear power plant or a city district heating system can have very large flow rates, while a pharmaceutical chemical plant may have very small flows, so we calibrate "small" using the nominal value (e.g. 15000 kg/s for the nuclear plant, 0.001 kg/s for the pharmaceutical plant). So the question boils down to: which nominal values of the flows in the connection set should be considered? If there are some large and some small, I guess the idea was to consider the smallest one, hence the min() operator. I think the intent was really min(nominal(mj.c.m_flow) for j in 1:N) as you wrote.

I'll open a ticket on the MLS to double check that. We should also check with @phanebohm what the backend is doing.

**Comment by casella** [comment]
> it simplifies some cases and otherwise just changes it to a normal `max` call. Since it doesn't have any information about the connection sets I guess it's hard to do much more than that.

Maybe the best way to handle that is that the frontend creates the functions (as required for Base Modelica) and passes them to the backend?

**Comment by henrikt-ma** [comment]
How would you go about creating a more sophisticated variant making use of nominals? Note that the design for `nominal` doesn't appear finished yet:
- Last item on roadmap is about `nominal` https://github.com/modelica/ModelicaSpecification/tree/MCP/0031/RationaleMCP/0031#base-modelica-01-this-mcp
- In _differences.md_, the section on `nominal` is just a TODO: https://github.com/modelica/ModelicaSpecification/blob/MCP/0031/RationaleMCP/0031/differences.md#the-nominal-attribute

Compare how guess values are handled using the implicitly declared guess value parameters of the form `guess('x')`.  If one could use `nominal('x')` to access (or modify) the nominal of `'x'`, one could then pass these as input to `positiveMax`.

**Comment by perost** [comment]
> How would you go about creating a more sophisticated variant making use of nominals? Note that the design for `nominal` doesn't appear finished yet:
> 
>     * Last item on roadmap is about `nominal` https://github.com/modelica/ModelicaSpecification/tree/MCP/0031/RationaleMCP/0031#base-modelica-01-this-mcp
> 
>     * In _differences.md_, the section on `nominal` is just a TODO: https://github.com/modelica/ModelicaSpecification/blob/MCP/0031/RationaleMCP/0031/differences.md#the-nominal-attribute
> 
> 
> Compare how guess values are handled using the implicitly declared guess value parameters of the form `guess('x')`. If one could use `nominal('x')` to access (or modify) the nominal of `'x'`, one could then pass these as input to `positiveMax`.

Currently I guess the only way is to use the binding expressions for the nominals directly, but this of course means that you get issues if the nominals of the variables are changed later. So having a `nominal` operator would definitely be helpful for this particular feature.

**Comment by casella** [comment]
See also #15449.
